File: Leetcode/400. Nth Digit/Solution.py
class Solution:
    def findNthDigit(self, n: int) -> int:
        length=1
        count=9
        start_num=1

        while n>length*count:
            n-=length*count
            length+=1
            count*=10
            start_num*=10

        target_number=start_num+(n-1)//length
        digit_index_in_num=(n-1)%length
        return int(str(target_number)[digit_index_in_num])
        

File: Leetcode/400. Nth Digit/Solution.cpp
#include <string>

class Solution {
public:
    int findNthDigit(int n) {
        long long digitLength = 1;
        long long count = 9;
        long long start = 1;

        while (n > digitLength * count) {
            n -= digitLength * count;
            digitLength++;
            count *= 10;
            start *= 10;
        }

        start += (n - 1) / digitLength;
        std::string number = std::to_string(start);
        return number[(n - 1) % digitLength] - '0';
    }
};
